### PR TITLE
feat: Add captcha service

### PR DESCRIPTION
**Pull request by BenjaminMichaelis**
This adds most of the logic needed for the captcha service. The UI implementation and locations needs to be flushed out a bit more before actually including, but this should cover most of the backend work.

### CODE REVIEW

File: EssentialCSharp.Web.Tests/Integration/CaptchaTests.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    public CaptchaTests(WebApplicationFactory<Program> factory)
+    {
+        _factory = factory;
+    }
+
+    [Fact]
+    public async Task CaptchaService_Verify_Success()
+    {
+        ICaptchaService? _CaptchaService = _factory.Services.GetService<ICaptchaService>();
+        Assert.NotNull(_CaptchaService);
```
Comment (Keboo) [inline EssentialCSharp.Web.Tests/Integration/CaptchaTests.cs:21]: nit: I would just use `GetRequiredService` which does the null check for you.
```suggestion
        ICaptchaService captchaService = _factory.Services.GetRequiredService<ICaptchaService>();

```

File: EssentialCSharp.Web/Controllers/HomeController.cs [hunk 2]
```diff
@@ -64,15 +69,36 @@ public IActionResult Announcements()
...
+        //{
+        //    _Logger.HomeControllerSuccessfulCaptchaResponse(Json(response));
+        //    return View();
+        //}
+        //else
+        //{
+        //    _Logger.HomeControllerSuccessfulCaptchaResponse(Json(response));
+
+        //    return RedirectToAction(nameof(Error), new { errorMessage = "Captcha Failed. Forbidden", statusCode = 403 });
+        //}
```
Comment (Keboo) [inline EssentialCSharp.Web/Controllers/HomeController.cs:98]: Don't we need this for the captcha to work?

File: EssentialCSharp.Web/Extensions/LoggerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+using System.Text;
+using Microsoft.AspNetCore.Mvc;
+
+namespace EssentialCSharp.Web.Extensions;
+
+internal static class LoggerExtensions
```
Comment (Keboo) [inline EssentialCSharp.Web/Extensions/LoggerExtensions.cs:6]: This works but we can leverage source generators to make this better.

```C#
internal static partial class LoggerExtensions
{
    [LoggerMessage(Level = LogLevel.Debug, EventId = 1, Message = "Successful captcha with response of: '{JsonResult}'")]
    public static partial void HomeControllerSuccessfulCaptchaResponse(
        this ILogger logger, JsonResult jsonResult);
}

File: EssentialCSharp.Web/Services/CaptchaService.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+﻿using Microsoft.Extensions.Options;
+
+namespace EssentialCSharp.Web.Services
```
Comment (Keboo) [inline EssentialCSharp.Web/Services/CaptchaService.cs:3]: Nit: File scoped namespace

File: EssentialCSharp.Web/Services/CaptchaService.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public async Task<HttpResponseMessage> Verify(string secret, string token)
+        {
+            // create post data
+            List<KeyValuePair<string, string>> postData = new()
+        {
+            new KeyValuePair<string, string>("secret", secret),
+            new KeyValuePair<string, string>("response", token),
+        };
```
Comment (Keboo) [inline EssentialCSharp.Web/Services/CaptchaService.cs:21]: nit: Whitespace
```suggestion
            {
                new KeyValuePair<string, string>("secret", secret),
                new KeyValuePair<string, string>("response", token),
            };

```

File: EssentialCSharp.Web/Services/CaptchaService.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        public async Task<HttpResponseMessage> Verify(string secret, string token, string remoteIp)
+        {
+            // create post data
+            List<KeyValuePair<string, string>> postData = new()
+        {
+            new KeyValuePair<string, string>("secret", secret),
+            new KeyValuePair<string, string>("response", token),
+            new KeyValuePair<string, string>("remoteip", remoteIp)
+        };
```
Comment (Keboo) [inline EssentialCSharp.Web/Services/CaptchaService.cs:34]: nit: Whitespace
```suggestion
            {
                new KeyValuePair<string, string>("secret", secret),
                new KeyValuePair<string, string>("response", token),
                new KeyValuePair<string, string>("remoteip", remoteIp)
            };

```

File: EssentialCSharp.Web/Services/ICaptchaService.cs [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿namespace EssentialCSharp.Web.Services
```
Comment (Keboo) [inline EssentialCSharp.Web/Services/ICaptchaService.cs:1]: File scoped namespace

File: EssentialCSharp.Web/Views/Home/Home.cshtml [hunk 8]
```diff
@@ -1,5 +1,13 @@
...
-    <div> 
+    <div>
+       @*  @using (Html.BeginForm())
+        {
+            <script src="https://js.hcaptcha.com/1/api.js" async defer></script>
+            <div class="form-group">
+                <div class="h-captcha" data-sitekey="b9235f58-3d8d-4394-ab8e-78b35a6d69c5"></div>
+            </div>
+            <button type="submit" class="btn btn-primary">Submit</button>
+        } *@
```
Comment (Keboo) [inline EssentialCSharp.Web/Views/Home/Home.cshtml:10]: Don't we need this to make it work?

File: EssentialCSharp.Web/Services/ICaptchaService.cs [hunk 9]
```diff
@@ -0,0 +1,8 @@
+﻿namespace EssentialCSharp.Web.Services
+{
+    public interface ICaptchaService
+    {
+        Task<HttpResponseMessage> Verify(string secret, string token, string remoteIp);
```
Comment (Keboo) [inline EssentialCSharp.Web/Services/ICaptchaService.cs:5]: Do we need this overload? It looks like we are only using the other one. 

Do we need both, or should remoteIp be an optional parameter (perhaps of type IPAddress?)?

File: EssentialCSharp.Web/Controllers/HomeController.cs (hunk 2 above)
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Controllers/HomeController.cs:98]: Yes... but I don't think this is where we want the captcha exactly yet, so I was going to hold off on actually hooking it into the website until the future when we could discuss it and flush it out more. I wanted to keep the code though just for reference but I can also delete for this PR. Thoughts @keboo?

Comment (Keboo) [reply EssentialCSharp.Web/Controllers/HomeController.cs:98]: That is fine. As long as we don't expect it working I am fine letting this merge.

File: EssentialCSharp.Web/Views/Home/Home.cshtml (hunk 8 above)
Comment (Keboo) [reply EssentialCSharp.Web/Views/Home/Home.cshtml:10]: As per the other thread, we are going to merge this and just holding this code as a reference.

File: EssentialCSharp.Web/Services/ICaptchaService.cs (hunk 9 above)
Comment (BenjaminMichaelis) [reply EssentialCSharp.Web/Services/ICaptchaService.cs:5]: Removed overload, just added a comment instead.

File: EssentialCSharp.Web/Services/CaptchaService.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    // https://docs.hcaptcha.com/#verify-the-user-response-server-side
+    public async Task<HttpResponseMessage> Verify(string secret, string token, string sitekey)
+    {
+        // create post data
+        List<KeyValuePair<string, string>> postData = new()
+    {
+        new KeyValuePair<string, string>("secret", secret),
+        new KeyValuePair<string, string>("response", token),
+        new KeyValuePair<string, string>("sitekey", sitekey)
+    };
```
Comment (Keboo) [inline EssentialCSharp.Web/Services/CaptchaService.cs:24]:
```suggestion
        {
            new KeyValuePair<string, string>("secret", secret),
            new KeyValuePair<string, string>("response", token),
            new KeyValuePair<string, string>("sitekey", sitekey)
        };

```

**Review by Keboo — APPROVED** [review]

**Pull request merged**